// oceanic.js@1.11.2/dist/lib/routes/Users.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Routes = tslib_1.__importStar(require("../util/Routes"));
const ExtendedUser_1 = tslib_1.__importDefault(require("../structures/ExtendedUser"));
/** Various methods for interacting with users. Located at {@link Client#rest | Client#rest}{@link RESTManager#users | .users}. */
class Users {
    _manager;
    constructor(manager) {
        this._manager = manager;
    }
    /** Alias for {@link REST/Channels#createDM | Channels#createDM}. */
    get createDM() {
        return this._manager.channels.createDM.bind(this._manager.channels);
    }
    /**
     * Edit the currently authenticated user.
     * @param options The options to edit with.
     * @caching This method **does not** cache its result.
     */
    async editSelf(options) {
        if (options.avatar) {
            options.avatar = this._manager.client.util._convertImage(options.avatar, "avatar");
        }
        if (options.banner) {
            options.banner = this._manager.client.util._convertImage(options.banner, "banner");
        }
        return this._manager.authRequest({
            method: "PATCH",
            path: Routes.USER("@me"),
            json: options
        }).then(data => new ExtendedUser_1.default(data, this._manager.client));
    }
    /**
     * Get a user.
     * @param userID the ID of the user
     * @caching This method **does** cache its result.
     * @caches {@link Client#users | Client#users}
     */
    async get(userID) {
        return this._manager.authRequest({
            method: "GET",
            path: Routes.USER(userID)
        }).then(data => this._manager.client.users.update(data));
    }
    /**
     * Leave a guild.
     * @param guildID The ID of the guild to leave.
     * @caching This method **does not** cache its result.
     */
    async leaveGuild(guildID) {
        await this._manager.authRequest({
            method: "DELETE",
            path: Routes.OAUTH_GUILD(guildID)
        });
    }
}
exports.default = Users;
