// oceanic.js@1.11.2/dist/lib/util/TypedCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module TypedCollection */
const Collection_1 = tslib_1.__importDefault(require("./Collection"));
const Base_1 = tslib_1.__importDefault(require("../structures/Base"));
/** This is an internal class, you should not use it in your projects. If you want a collection type for your own projects, look at {@link Collection}. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
class TypedCollection extends Collection_1.default {
    _baseObject;
    extraOptions;
    limit;
    constructor(baseObject, client, limit = Infinity, extraOptions) {
        super();
        if (!(baseObject.prototype instanceof Base_1.default)) {
            throw new TypeError("baseObject must be a class that extends Base.");
        }
        this._baseObject = baseObject;
        this.limit = limit;
        this.extraOptions = {
            construct: extraOptions?.construct ?? ((data, ...extra) => new baseObject(data, client, ...extra)),
            delete: extraOptions?.delete ?? (() => { })
        };
    }
    /** @hidden */
    add(value) {
        if ("id" in value) {
            if (this.limit === 0) {
                return value;
            }
            this.set(value.id, value);
            if (this.limit && this.size > this.limit) {
                const iter = this.keys();
                while (this.size > this.limit) {
                    this.delete(iter.next().value);
                }
            }
            return value;
        }
        else {
            const err = new Error(`${this.constructor.name}#add: value must have an id property`);
            Object.defineProperty(err, "_object", { value });
            throw err;
        }
    }
    clear() {
        for (const key of this.keys()) {
            this.extraOptions.delete(key);
        }
        super.clear();
    }
    delete(key) {
        this.extraOptions.delete(key);
        return super.delete(key);
    }
    /** @hidden */
    update(value, ...extra) {
        if (value instanceof this._baseObject) {
            if ("update" in value) {
                value["update"].call(value, value);
            }
            return value;
        }
        // if the object does not have a direct id, we're forced to construct a whole new object
        let item = "id" in value && value.id ? this.get(value.id) : undefined;
        if (!item) {
            item = this.add(this.extraOptions.construct(value, ...extra));
        }
        else if ("update" in item) {
            item["update"].call(item, value);
        }
        return item;
    }
}
exports.default = TypedCollection;
