// oceanic.js@1.11.2/dist/lib/util/interactions/SelectMenuValuesWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** @module SelectMenuValuesWrapper */
const Errors_1 = require("../Errors");
const Constants_1 = require("../../Constants");
/** A wrapper for select menu data. */
class SelectMenuValuesWrapper {
    /** The raw received values. */
    raw;
    /** The resolved data for this instance. */
    resolved;
    constructor(resolved, values) {
        this.resolved = resolved;
        this.raw = values;
    }
    /**
     * Get the selected channels.
     *
     * If `ensurePresent` is false, channels that aren't in resolved will be ignored.
     * @param ensurePresent If true, an error will be thrown if any value cannot be mapped to a channel.
     */
    getChannels(ensurePresent) {
        return this.raw.map(id => {
            const ch = this.resolved.channels.get(id);
            if (!ch && ensurePresent) {
                throw new Errors_1.WrapperError(`Failed to find channel in resolved data: ${id}`);
            }
            return ch;
        }).filter(Boolean);
    }
    /**
     * Get the complete version of the selected channels. This will only succeed if the channel is cached. If the channel is private and isn't cached, an `InteractionResolvedChannel` instance will still be returned.
     *
     * If `ensurePresent` is false, channels that aren't in resolved will be ignored.
     * @param ensurePresent If true, an error will be thrown if any value cannot be mapped to a channel.
     */
    getCompleteChannels(ensurePresent) {
        return this.raw.map(id => {
            const ch = this.resolved.channels.get(id);
            if (ch && ch.type === Constants_1.ChannelTypes.DM) {
                return ch?.completeChannel ?? ch;
            }
            if (!ch && ensurePresent) {
                throw new Errors_1.WrapperError(`Failed to find channel in resolved data: ${id}`);
            }
            return ch;
        }).filter(Boolean);
    }
    /**
     * Get the selected members.
     *
     * If `ensurePresent` is false, members that aren't in resolved will be ignored.
     * @param ensurePresent If true, an error will be thrown if any value cannot be mapped to a member.
     */
    getMembers(ensurePresent) {
        return this.raw.map(id => {
            const member = this.resolved.members.get(id);
            if (!member && ensurePresent) {
                throw new Errors_1.WrapperError(`Failed to find member in resolved data: ${id}`);
            }
            return member;
        }).filter(Boolean);
    }
    /**
     * Get the selected mentionables. (users, roles)
     *
     * If `ensurePresent` is false, mentionables that aren't in resolved will be ignored.
     * @param ensurePresent If true, an error will be thrown if any value cannot be mapped to a user, or role.
     */
    getMentionables(ensurePresent) {
        const res = [];
        for (const id of this.raw) {
            const role = this.resolved.roles.get(id);
            const user = this.resolved.users.get(id);
            if ((!role && !user)) {
                if (ensurePresent) {
                    throw new Errors_1.WrapperError(`Failed to find mentionable in resolved data: ${id}`);
                }
            }
            else {
                res.push((role ?? user));
            }
        }
        return res;
    }
    /**
     * Get the selected roles.
     *
     * If `ensurePresent` is false, roles that aren't in resolved will be ignored.
     * @param ensurePresent If true, an error will be thrown if any value cannot be mapped to a role.
     */
    getRoles(ensurePresent) {
        return this.raw.map(id => {
            const role = this.resolved.roles.get(id);
            if (!role && ensurePresent) {
                throw new Errors_1.WrapperError(`Failed to find role in resolved data: ${id}`);
            }
            return role;
        }).filter(Boolean);
    }
    /**
     * Get the selected string values. This cannot fail.
     */
    getStrings() {
        return this.raw;
    }
    /**
     * Get the selected users.
     *
     * If `ensurePresent` is false, users that aren't in resolved will be ignored.
     * @param ensurePresent If true, an error will be thrown if any value cannot be mapped to a user.
     */
    getUsers(ensurePresent) {
        return this.raw.map(id => {
            const user = this.resolved.users.get(id);
            if (!user && ensurePresent) {
                throw new Errors_1.WrapperError(`Failed to find user in resolved data: ${id}`);
            }
            return user;
        }).filter(Boolean);
    }
}
exports.default = SelectMenuValuesWrapper;
