// oceanic.js@1.11.2/dist/lib/structures/Attachment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module Attachment */
const Base_1 = tslib_1.__importDefault(require("./Base"));
/** Represents a file attachment. */
class Attachment extends Base_1.default {
    /** The mime type of this attachment. */
    contentType;
    /** The description of this attachment. */
    description;
    /** The duration of the attached audio file, if voice message. */
    durationSecs;
    /** If this attachment is ephemeral. Ephemeral attachments will be removed after a set period of time. */
    ephemeral;
    /** The filename of this attachment. */
    filename;
    /** The {@link Constants~AttachmentFlags | Attachment Flags } of this image. */
    flags;
    /** The height of this attachment, if an image. */
    height;
    /** A proxied url of this attachment. */
    proxyURL;
    /** The size of this attachment. */
    size;
    /** The title of this attachment. */
    title;
    /** The source url of this attachment. */
    url;
    /** Base64 encoded bytearray representing a sampled waveform for voice messages. */
    waveform;
    /** The width of this attachment, if an image. */
    width;
    constructor(data, client) {
        super(data.id, client);
        this.contentType = data.content_type;
        this.description = data.description;
        this.durationSecs = data.duration_secs;
        this.ephemeral = data.ephemeral;
        this.filename = data.filename;
        this.flags = data.flags ?? 0;
        this.height = data.height;
        this.proxyURL = data.proxy_url;
        this.size = data.size;
        this.url = data.url;
        this.waveform = data.waveform;
        this.width = data.width;
    }
    toJSON() {
        return {
            ...super.toJSON(),
            contentType: this.contentType,
            description: this.description,
            ephemeral: this.ephemeral,
            filename: this.filename,
            flags: this.flags,
            height: this.height,
            proxyURL: this.proxyURL,
            size: this.size,
            title: this.title,
            url: this.url,
            width: this.width
        };
    }
}
exports.default = Attachment;
