// oceanic.js@1.11.2/dist/lib/structures/BaseEntitlement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Base_1 = tslib_1.__importDefault(require("./Base"));
/** Represents a base entitlement. See {@link TestEntitlement | TestEntitlement} and {@link Entitlement | Entitlement}. */
class BaseEntitlement extends Base_1.default {
    applicationID;
    consumed;
    deleted;
    giftCodeFlags;
    guildID;
    promotionID;
    skuID;
    type;
    userID;
    constructor(data, client) {
        super(data.id, client);
        this.applicationID = data.application_id;
        this.consumed = data.consumed;
        this.deleted = data.deleted;
        this.giftCodeFlags = data.gift_code_flags;
        this.guildID = data.guild_id;
        this.promotionID = data.promotion_id;
        this.skuID = data.sku_id;
        this.type = data.type;
        this.userID = data.user_id;
    }
    /** Mark this entitlement as consumed. */
    async consume() {
        return this.client.rest.applications.consumeEntitlement(this.applicationID, this.id);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            applicationID: this.applicationID,
            consumed: this.consumed,
            deleted: this.deleted,
            giftCodeFlags: this.giftCodeFlags,
            guildID: this.guildID,
            promotionID: this.promotionID,
            skuID: this.skuID,
            type: this.type,
            userID: this.userID
        };
    }
}
exports.default = BaseEntitlement;
