// oceanic.js@1.11.2/dist/lib/structures/Clan.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module Clan */
const Routes = tslib_1.__importStar(require("../util/Routes"));
/** Represents an invite. */
class Clan {
    /** The badge hash of this clan. */
    badge;
    client;
    identityEnabled;
    identityGuildID;
    /** The tag of this clan, shown beside messages. */
    tag;
    constructor(data, client) {
        Object.defineProperty(this, "client", {
            value: client,
            enumerable: false,
            writable: false,
            configurable: false
        });
        this.badge = data.badge;
        this.identityEnabled = data.identity_enabled;
        this.identityGuildID = data.identity_guild_id;
        this.tag = data.tag;
        this.update(data);
    }
    update(data) {
        if (data.badge !== undefined) {
            this.badge = data.badge;
        }
        if (data.identity_enabled !== undefined) {
            this.identityEnabled = data.identity_enabled;
        }
        if (data.identity_guild_id !== undefined) {
            this.identityGuildID = data.identity_guild_id;
        }
        if (data.tag !== undefined) {
            this.tag = data.tag;
        }
    }
    /**
     * The url of this guild's icon.
     * @param format The format the url should be.
     * @param size The dimensions of the image.
     */
    badgeURL(format, size) {
        return this.client.util.formatImage(Routes.CLAN_ICON(this.identityGuildID, this.badge), format, size);
    }
    toJSON() {
        return {
            badge: this.badge,
            identityEnabled: this.identityEnabled,
            identityGuildID: this.identityGuildID,
            tag: this.tag
        };
    }
}
exports.default = Clan;
