// oceanic.js@1.11.2/dist/lib/structures/SKU.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Base_1 = tslib_1.__importDefault(require("./Base"));
class SKU extends Base_1.default {
    accessType; // undocumented
    applicationID;
    dependentSKUID;
    features; // undocumented
    /** The flags for this SKU. See {@link Constants~SKUFlags | SKUFlags}. */
    flags;
    manifestLabels; // undocumented
    name;
    releaseDate; // undocumented
    showAgeGate;
    slug;
    type;
    constructor(data, client) {
        super(data.id, client);
        this.accessType = data.access_type;
        this.applicationID = data.application_id;
        this.dependentSKUID = data.dependent_sku_id;
        this.features = data.features;
        this.flags = data.flags;
        this.manifestLabels = data.manifest_labels;
        this.name = data.name;
        this.releaseDate = data.release_date;
        this.showAgeGate = data.show_age_gate;
        this.slug = data.slug;
        this.type = data.type;
    }
    /**
     * Create a test entitlement for this SKU.
     * @param ownerType The type of the owner to create the entitlement for.
     * @param ownerID The ID of the owner to create the entitlement for.
     */
    async createTestEntitlement(ownerType, ownerID) {
        return this.client.rest.applications.createTestEntitlement(this.applicationID, {
            ownerID,
            ownerType,
            skuID: this.id
        });
    }
    /**
     * Get the entitlements for this SKU.
     * @param options The options for getting the entitlements.
     */
    async getEntitlements(options) {
        return this.client.rest.applications.getEntitlements(this.applicationID, { skuIDs: [this.id], ...options });
    }
}
exports.default = SKU;
