// oceanic.js@1.11.2/dist/lib/structures/TextableVoiceChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module TextableVoiceChannel */
const Member_1 = tslib_1.__importDefault(require("./Member"));
const TextableChannel_1 = tslib_1.__importDefault(require("./TextableChannel"));
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a textable voice channel. */
class TextableVoiceChannel extends TextableChannel_1.default {
    /** The bitrate of the stage channel. */
    bitrate;
    /** The id of the voice region of the channel, `null` is automatic. */
    rtcRegion;
    /** The maximum number of members in this voice channel, `0` is unlimited. */
    userLimit;
    /** The [video quality mode](https://discord.com/developers/docs/resources/channel#channel-object-video-quality-modes) of this channel. */
    videoQualityMode;
    voiceMembers;
    constructor(data, client) {
        super(data, client);
        this.bitrate = data.bitrate;
        this.rtcRegion = data.rtc_region;
        this.userLimit = data.user_limit;
        this.videoQualityMode = data.video_quality_mode;
        this.voiceMembers = new TypedCollection_1.default(Member_1.default, client, this.client.util._getLimit("voiceMembers", this.id));
        this.update(data);
    }
    update(data) {
        super.update(data);
        if (data.bitrate !== undefined) {
            this.bitrate = data.bitrate;
        }
        if (data.rtc_region !== undefined) {
            this.rtcRegion = data.rtc_region;
        }
        if (data.user_limit !== undefined) {
            this.userLimit = data.user_limit;
        }
        if (data.video_quality_mode !== undefined) {
            this.videoQualityMode = data.video_quality_mode;
        }
    }
    get parent() {
        return super.parent;
    }
    /** The voice states related to this channel. */
    get voiceStates() {
        return this["_cachedGuild"]?.voiceStates.filter(state => state.channelID === this.id) ?? [];
    }
    /**
     * Join this stage channel.
     * @param options The options to join the channel with.
     */
    join(options) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument
        return this.client.joinVoiceChannel({
            ...options,
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            voiceAdapterCreator: this.guild.voiceAdapterCreator,
            guildID: this.guildID,
            channelID: this.id
        });
    }
    /** Leave this stage channel. */
    leave() {
        return this.client.leaveVoiceChannel(this.guildID);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            bitrate: this.bitrate,
            rtcRegion: this.rtcRegion,
            type: this.type,
            userLimit: this.userLimit,
            videoQualityMode: this.videoQualityMode,
            voiceMembers: this.voiceMembers.map(member => member.id)
        };
    }
}
exports.default = TextableVoiceChannel;
