// oceanic.js@1.11.2/dist/lib/structures/VoiceState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module VoiceState */
const Base_1 = tslib_1.__importDefault(require("./Base"));
const Errors_1 = require("../util/Errors");
/** Represents a guild member's voice state. */
class VoiceState extends Base_1.default {
    _cachedChannel;
    _cachedGuild;
    _cachedMember;
    _cachedUser;
    /** The ID of the channel the user is connected to. */
    channelID;
    /** If the associated member is deafened. */
    deaf;
    /** The ID of the guild this voice state is a part of. */
    guildID;
    /** If the associated member is muted. */
    mute;
    /** The time at which the associated member requested to speak. */
    requestToSpeakTimestamp;
    /** If the associated member is self deafened. */
    selfDeaf;
    /** If the associated member is self muted. */
    selfMute;
    /** If the associated member is streaming. */
    selfStream;
    /** If the associated member is has their camera on. */
    selfVideo;
    /** The id of the associated member's voice session. */
    sessionID;
    /** If the associated member is suppressed. */
    suppress;
    /** The ID of the user associated with this voice state. */
    userID;
    constructor(data, client) {
        super(data.user_id, client);
        this.channelID = data.channel_id;
        this.deaf = false;
        this.guildID = data.guild_id;
        this.mute = false;
        this.requestToSpeakTimestamp = null;
        this.selfDeaf = false;
        this.selfMute = false;
        this.selfStream = false;
        this.selfVideo = false;
        this.suppress = false;
        this.userID = this.id;
        this.update(data);
    }
    update(data) {
        if (data.channel_id !== undefined) {
            this.channelID = data.channel_id;
            this._cachedChannel = null;
        }
        if (data.deaf !== undefined) {
            this.deaf = data.deaf;
        }
        if (data.guild_id !== undefined) {
            this.guildID = data.guild_id;
        }
        if (data.member !== undefined) {
            this._cachedMember = this.client.util.updateMember(data.guild_id, this.id, data.member);
        }
        if (data.mute !== undefined) {
            this.mute = data.mute;
        }
        if (data.request_to_speak_timestamp !== undefined) {
            this.requestToSpeakTimestamp = data.request_to_speak_timestamp === null ? null : new Date(data.request_to_speak_timestamp);
        }
        if (data.self_deaf !== undefined) {
            this.selfDeaf = data.self_deaf;
        }
        if (data.self_mute !== undefined) {
            this.selfMute = data.self_mute;
        }
        if (data.self_stream !== undefined) {
            this.selfStream = data.self_stream;
        }
        if (data.self_video !== undefined) {
            this.selfVideo = data.self_video;
        }
        if (data.suppress !== undefined) {
            this.suppress = data.suppress;
        }
        if (data.user_id !== undefined) {
            this.userID = data.user_id;
        }
    }
    /** The channel the user is connected to. */
    get channel() {
        if (this.channelID !== null) {
            return this._cachedChannel ??= this.client.getChannel(this.channelID);
        }
        return (this._cachedChannel ??= null);
    }
    /** The guild this voice state is a part of. This will throw an error if the guild is not cached. */
    get guild() {
        this._cachedGuild ??= this.client.guilds.get(this.guildID);
        if (!this._cachedGuild) {
            if (this.client.options.restMode) {
                throw new Errors_1.UncachedError(`${this.constructor.name}#guild is not present when rest mode is enabled.`);
            }
            if (!this.client.shards.connected) {
                throw new Errors_1.UncachedError(`${this.constructor.name}#guild is not present without a gateway connection.`);
            }
            throw new Errors_1.UncachedError(`${this.constructor.name}#guild is not present.`);
        }
        return this._cachedGuild;
    }
    /** The member associated with this voice state. */
    get member() {
        return this._cachedMember ?? (this._cachedMember = this["_cachedGuild"]?.members.get(this.userID));
    }
    /** TThe user associated with this voice state. */
    get user() {
        return this._cachedUser ?? (this._cachedUser = this.client.users.get(this.userID));
    }
    toJSON() {
        return {
            ...super.toJSON(),
            channelID: this.channelID,
            deaf: this.deaf,
            guildID: this.guildID ?? undefined,
            member: this.member?.toJSON(),
            mute: this.mute,
            requestToSpeakTimestamp: this.requestToSpeakTimestamp ? this.requestToSpeakTimestamp.getTime() : null,
            selfDeaf: this.selfDeaf,
            selfMute: this.selfMute,
            selfStream: this.selfStream,
            selfVideo: this.selfVideo,
            sessionID: this.sessionID,
            suppress: this.suppress,
            user: this.user?.toJSON()
        };
    }
}
exports.default = VoiceState;
